File: src/01_introduction_to_algorithms/ts/binary_search.ts
class BinarySearch {
  private low: number = 0;
  private high: number = 0;

  search_iterative(item: number, list: number[]): number | null {
    this.high = list.length - 1;
    while (this.low <= this.high) {
      // check the middle element
      const mid: number = Math.round((this.low + this.high) / 2);
      const guess: number = list[mid];
      // Found the item.
      if (guess == item) {
        return mid;
      }
      // The guess was too high.
      else if (guess > item) {
        this.high = mid - 1;
      } else {
        // The guess was too low.
        this.low = mid + 1;
      }
    }
    // Item doesn't exist
    return null;
  }

  search_recursive(item: number, list: number[], low: number = 0, high?: number): number | null {
    this.low = low;
    high ? (this.high = high) : (this.high = list.length - 1);

    // Check base case
    if (this.high >= this.low) {
      // check the middle element
      const mid: number = Math.round((this.low + this.high) / 2);
      const guess: number = list[mid];

      // Found the item.
      if (guess == item) {
        return mid;
      }
      // If element is smaller than mid, then it can only
      // be present in left subarray
      else if (guess > item) {
        return this.search_recursive(item, list, this.low, mid - 1);
      } else {
        // The guess was too low.
        return this.search_recursive(item, list, mid + 1, this.high);
      }
    } else {
      // Item doesn't exist
      return null;
    }
  }
}

const bs = new BinarySearch();
const my_list = [1, 3, 5, 7, 9];

// console.log(bs.search_iterative(9, my_list)); // index 4
console.log(bs.search_recursive(9, my_list)); // index 4

File: src/01_introduction_to_algorithms/python/binary_search.py
class BinarySearch():

    def search_iterative(self, arr: list[int], item: int) -> (int | None ):
        # low and high keep track of which part of the list you'll search in.
        low = 0
        high = len(arr) - 1

        # While you haven't narrowed it down to one element ...
        while low <= high:
            # ... check the middle element
            mid = (low + high) // 2
            guess = arr[mid]
            # Found the item.
            if guess == item:
                return mid
            # The guess was too high.
            if guess > item:
                high = mid - 1
            # The guess was too low.
            else:
                low = mid + 1

        # Item doesn't exist
        return None

    def search_recursive(self, arr: list[int], low: int, high: int, item: int) -> (int | None ):
        # Check base case
        if high >= low:
            mid = (high + low) // 2
            guess = arr[mid]

            # If element is present at the middle itself
            if guess == item:
                return mid

            # If element is smaller than mid, then it can only
            # be present in left subarray
            elif guess > item:
                return self.search_recursive(arr, low, mid - 1, item)

            # Else the element can only be present in right subarray
            else:
                return self.search_recursive(arr, mid + 1, high, item)

        else:
            # Element is not present in the array
            return None


if __name__ == "__main__":
    # We must initialize the class to use the methods of this class
    bs = BinarySearch()
    my_list = [1, 3, 5, 7, 9]

    print(bs.search_iterative(my_list, 9))  # => retorna a posição no array

    # 'None' means null in Python. We use to indicate that the
    # item wasn't found.
    print(bs.search_iterative(my_list, 10))  # => None
